src/fsevents/fsevents_stubs.c

```diff
@@ -0,0 +1,221 @@
+  CAMLparam0();
+  CAMLlocal5(v_events_xs, v_events_x, v_flags, v_id, v_event);
+  CAMLlocal1(v_path);
+  v_events_xs = Val_emptylist;
+
+  // we iterate over the events backwards to avoid reversing the list in the end
+  size_t j = 0;
+  size_t i = numEvents - 1;
+  for (; j < numEvents; j++, i = numEvents - j - 1) {
+    FSEventStreamEventFlags flags = eventFlags[i];
+    if (!(interesting_flags & flags)) {
+      continue;
+    }
+    CFDictionaryRef details = CFArrayGetValueAtIndex(eventPaths, i);
+    CFStringRef cf_path =
+        CFDictionaryGetValue(details, kFSEventStreamEventExtendedDataPathKey);
+    CFIndex len = CFStringGetLength(cf_path);
+    CFIndex byte_len;
+    CFIndex res =
+        CFStringGetBytes(cf_path, CFRangeMake(0, len), kCFStringEncodingUTF8, 0,
+                         0, NULL, 0, &byte_len);
+    if (res == 0) {
+      // TODO
+      exit(456);
+    }
+    v_path = caml_alloc_string(byte_len);
+    unsigned char *p = Bytes_val(v_path);
+    res = CFStringGetBytes(cf_path, CFRangeMake(0, len), kCFStringEncodingUTF8,
+                           0, 0, (UInt8 *)p, byte_len, NULL);
+    if (res == 0) {
+      // TODO
+      exit(123);
+    }
+
+    v_event = caml_alloc(3, 0);
+    v_id = integers_copy_uint64(eventIds[i]);
+    v_flags = integers_copy_uint32(flags); // TODO use ocaml int
+    Store_field(v_event, 0, v_path);
+    Store_field(v_event, 1, v_id);
+    Store_field(v_event, 2, v_flags);
+
+    v_events_x = caml_alloc(2, 0);
+    Store_field(v_events_x, 0, v_event);
+    Store_field(v_events_x, 1, v_events_xs);
+    v_events_xs = v_events_x;
+  }
+  caml_callback2(t->v_callback, caml_copy_nativeint((intnat)t), v_events_xs);
+  CAMLreturn0;
+}
+
+CAMLprim value dune_fsevents_create(value v_paths, value v_latency,
+                                    value v_callback) {
+  CAMLparam3(v_paths, v_latency, v_callback);
+  CAMLlocal1(path);
+
+  CFMutableArrayRef paths =
+      CFArrayCreateMutable(kCFAllocatorDefault, 0, &kCFTypeArrayCallBacks);
+
+  int i = 0;
+  while (v_paths != Val_emptylist) {
+    path = Field(v_paths, 0);
+    CFStringRef s = CFStringCreateWithCString(
+        kCFAllocatorDefault, String_val(path), kCFStringEncodingUTF8);
+    CFArraySetValueAtIndex(paths, i, s);
+    v_paths = Field(v_paths, 1);
+    i++;
+  }
+
+  const FSEventStreamEventFlags flags =
+      kFSEventStreamCreateFlagNoDefer |
+      kFSEventStreamCreateFlagUseExtendedData |
+      kFSEventStreamCreateFlagUseCFTypes | kFSEventStreamCreateFlagFileEvents;
+
+  dune_fsevents_t *t;
+  t = malloc(sizeof(dune_fsevents_t));
+
+  FSEventStreamContext context = {0, (void *)t, NULL /*retain*/,
+                                  NULL /*release*/, NULL};
+  FSEventStreamRef stream = FSEventStreamCreate(
+      kCFAllocatorDefault, (FSEventStreamCallback)&dune_fsevents_callback,
+      &context, paths, kFSEventStreamEventIdSinceNow, Double_val(v_latency),
+      flags);
+  CFRelease(paths);
+  caml_register_global_root(&v_callback);
```

Comment:
Is this correct? You are registering as a root the address of the *local* variable `v_callback` which will cease to exist once the function returns. Shouldn't it be `caml_register_global_root(&t->v_callback)`?

Reply:
Indeed. Thanks.